pulser/simulation/simresults.py

```diff
@@ -162,22 +355,23 @@ def sample_final_state(self, meas_basis=None, N_samples=1000):
             ValueError: If trying to sample without a defined 'meas_basis' in
                 the arguments when the original sequence is not measured.
         """
-        if meas_basis is None:
-            if self._meas_basis is None:
-                raise ValueError(
-                    "Can't accept an undefined measurement basis because the "
-                    "original sequence has no measurement."
-                    )
+        if not meas_basis:
             meas_basis = self._meas_basis
-
         if meas_basis not in {'ground-rydberg', 'digital'}:
             raise ValueError(
-                "'meas_basis' can only be 'ground-rydberg' or 'digital'."
+                "`meas_basis` can only be 'ground-rydberg' or 'digital'."
                 )
 
         N = self._size
         self.N_samples = N_samples
-        probs = np.abs(self._states[-1].full())**2
+        final_state = self._states[t]
+        # Case of a density matrix
+        # Don't take the modulus square in this case !
+        # if final_state.type != "ket":
```

Comment:
In the case of the `'dephasing'` noise, the Lindblad Master Equation will be called, so there will be non-zero off-diagonal elements in the resulting density matrix. In that case, the diagonal should be taken into account (commented lines). This corresponds to the projective measurements in the computational basis.

Reply:
Indeed, now uncommented.